File: src/main/kotlin/engine/knowledge/TeamKnowledge.kt
package eelst.ilike.engine.knowledge

import eelst.ilike.game.GloballyAvailableGameData
import eelst.ilike.game.entity.HanabiCard
import eelst.ilike.game.entity.action.ClueAction
import eelst.ilike.game.entity.action.DrawAction
import eelst.ilike.game.entity.player.PlayerMetadata

interface TeamKnowledge {
    /**
     * Get what the team knows about the slot with the given [slotIndex] in the hand of the player with the given
     * [playerIndex]
     */
    fun getSlotKnowledge(playerIndex: Int, slotIndex: Int): SlotKnowledge

    /**
     * @return the updated [TeamKnowledge] after a new (unspecified) card is drawn and added to the player's slot 1
     */
    fun getAfterDraw(drawAction: DrawAction, globallyAvailableGameData: GloballyAvailableGameData): TeamKnowledge

    /**
     * @return the updated [TeamKnowledge] after the specified [card] is drawn and added to the player's slot 1
     */
    fun getAfterDrawing(
        drawAction: DrawAction,
        card: HanabiCard,
        globallyAvailableGameData: GloballyAvailableGameData,
        ): TeamKnowledge

    /**
     * @return this [TeamKnowledge] after the slot with index [slotIndex] of the player with metadata [playerMetadata]
     * is removed from the player's hand (by playing it or discarding it
     */
    fun withoutSlot(slotIndex: Int, playerMetadata: PlayerMetadata): TeamKnowledge

    /**
     * @return the updated [TeamKnowledge] after a player gives a clue
     */
    fun getAfterClueGiven(
        clueAction: ClueAction,
        touchedSlotsIndexes: Collection<Int>,
        globallyAvailableGameData: GloballyAvailableGameData,
    ): TeamKnowledge

    /**
     * @return all the cards which the player with the given [playerIndex] can  see in the hands of each players
     * (including the own hand)
     */
    fun getCardsVisibleOnHands(playerIndex: Int): Collection<HanabiCard>

    /**
     * @return a new [TeamKnowledge] combining the knowledge contained in this object and [other]
     */
    fun getMergedWith(other: TeamKnowledge): TeamKnowledge
}

File: src/engine/knowledge/team_knowledge.rs
use crate::engine::knowledge::player_knowledge::PlayerKnowledge;
use crate::game::MAX_PLAYERS_IN_GAME;
use crate::game::card::{CardDeckIndex, VariantCardId};

/// Collective representation of what all players on the team know.
///
/// Stores one [`PlayerKnowledge`] per player in a fixed-size array
/// (sized to `MAX_PLAYERS_IN_GAME`) for stack-friendly, clone-cheap search nodes.
#[derive(Clone, Debug, Eq, PartialEq)]
pub struct TeamKnowledge {
    player_knowledge: [PlayerKnowledge; MAX_PLAYERS_IN_GAME],
    num_players: usize,
}

impl TeamKnowledge {
    #[must_use]
    pub fn new(num_players: usize) -> Self {
        let mut player_knowledge = std::array::from_fn(|_| PlayerKnowledge::empty());
        for i in 0..num_players {
            player_knowledge[i] = PlayerKnowledge::new(i);
        }
        TeamKnowledge {
            player_knowledge,
            num_players,
        }
    }

    /// Get a reference to a specific player's knowledge.
    #[must_use]
    pub fn player(&self, player_index: usize) -> &PlayerKnowledge {
        debug_assert!(player_index < self.num_players);
        &self.player_knowledge[player_index]
    }

    /// Get a mutable reference to a specific player's knowledge.
    pub fn player_mut(&mut self, player_index: usize) -> &mut PlayerKnowledge {
        debug_assert!(player_index < self.num_players);
        &mut self.player_knowledge[player_index]
    }

    /// Update all players' knowledge when a new card is drawn by one of the players.
    ///
    /// Players other than the drawer can see the card and have their empathy updated.
    /// The drawer cannot see the card but knows it is now in their hand.
    pub fn update_with_card_drawn(
        &mut self,
        player_index: usize,
        card_position_in_starting_deck: CardDeckIndex,
        card_id: VariantCardId,
    ) {
        for i in 0..self.num_players {
            if i != player_index {
                self.player_knowledge[i]
                    .update_with_revealed_card(card_position_in_starting_deck, card_id);
            }
        }
        self.player_knowledge[player_index].own_hand |= 1 << card_position_in_starting_deck;
    }

    /// Number of active players.
    #[must_use]
    pub fn num_players(&self) -> usize {
        self.num_players
    }
}
